File: Python/2016/01.py
from lib import *

input = read_input(2016, 1)

x = y = 0
dx, dy = 1, 0
for e in input.split(", "):
    l = int(e[1:])
    if e[0] == "L":
        dx, dy = dy, -dx
    else:
        dx, dy = -dy, dx

    x += l * dx
    y += l * dy

print(abs(x) + abs(y))


x = y = 0
dx, dy = 1, 0
seen = {(0, 0)}
for e in input.split(", "):
    l = int(e[1:])
    if e[0] == "L":
        dx, dy = dy, -dx
    else:
        dx, dy = -dy, dx

    for i in range(l):
        x += dx
        y += dy
        if (x, y) in seen:
            print(abs(x) + abs(y))
            break
        seen.add((x, y))
    else:
        continue
    break

File: Haskell/2024/01.hs
import Data.Function
import Data.List
import Lib

type Input = [(Int, Int)]

main :: IO ()
main = aoc 2024 1 setup solve1 solve2 ["1"]

solve1 :: Input -> Int
solve1 = sum . map abs . fork (on (zipWith (-)) sort) (map fst) (map snd)

solve2 :: Input -> Int
solve2 inp = sum $ map ((*) <*> (`count` right)) left
  where
    (left, right) = fork (,) (map fst) (map snd) inp
    count a = length . filter (== a)

setup :: String -> Input
setup = map (listToTuple . map read . words) . lines
